ui/src/chat-with-kg/App.tsx

```diff
@@ -1,35 +1,100 @@
-import { useState } from "react";
+import { useCallback, useEffect, useState } from "react";
 import ChatContainer from "./ChatContainer";
 import type { ChatMessageObject } from "./ChatMessage";
 import ChatInput from "./ChatInput";
-import { fetchQuestionAnswer } from "./utils/fetch-utils";
+import useWebSocket, { ReadyState } from "react-use-websocket";
+import {
+  ConversationState,
+  WebSocketRequest,
+  WebSocketResponse,
+} from "./types/websocketTypes";
 
 const SEND_REQUESTS = true;
 
-const chatMessageObjects: ChatMessageObject[] = SEND_REQUESTS ? [] : [
-  {
-    id: 0,
-    type: "input",
-    sender: "self",
-    message:
-      "This is the first message which has decently long text and would denote something typed by the user",
-  },
-  {
-    id: 1,
-    type: "text",
-    sender: "bot",
-    message:
-      "And here is another message which would denote a response from the server, which for now will only be text",
-  },
-];
+const chatMessageObjects: ChatMessageObject[] = SEND_REQUESTS
+  ? []
+  : [
+      {
+        id: 0,
+        type: "input",
+        sender: "self",
+        message:
+          "This is the first message which has decently long text and would denote something typed by the user",
+      },
+      {
+        id: 1,
+        type: "text",
+        sender: "bot",
+        message:
+          "And here is another message which would denote a response from the server, which for now will only be text",
+      },
+    ];
+
+const URL =
+  import.meta.env.VITE_KG_CHAT_BACKEND_ENDPOINT ??
+  "ws://localhost:7860/text2text";
 
 function App() {
-  const [chatMessages, setChatMesages] = useState(chatMessageObjects);
-  const [loading, setLoading] = useState(false);
+  const [chatMessages, setChatMessages] = useState(chatMessageObjects);
+  const [conversationState, setConversationState] =
+    useState<ConversationState>("ready");
+  const { sendJsonMessage, lastMessage, readyState } = useWebSocket(URL);
+
+  useEffect(() => {
+    if (!lastMessage) {
+      return;
+    }
+
+    const websocketResponse = JSON.parse(lastMessage.data) as WebSocketResponse;
+    console.log(websocketResponse);
+
+    if (websocketResponse.type === "debug") {
+      console.log(websocketResponse.detail);
+    } else if (websocketResponse.type === "error") {
+      setConversationState("error");
+      console.error(websocketResponse.detail);
+    } else if (websocketResponse.type === "start") {
+      setConversationState("streaming");
+
+      setChatMessages((chatMessages) => [
+        ...chatMessages,
+        {
+          id: chatMessages.length,
+          type: "text",
+          sender: "bot",
+          message: "",
```

Comment:
might be smart to explicitly flag the last message as being incomplete or something, even though message being an empty string could technically be used to denote that...